# **Exploratory Data Analysis of Historical Stock Market Data**

## **Introduction**:

This project focuses on the Exploratory Data Analysis (EDA) of historical stock market data from 1970 to 2018. We will use two datasets. The historical_stocks.csv dataset contains information about stocks, including ticker, exchange, company name, sector, and industry. The historical_stock_prices.csv dataset contains historical stock data, including ticker, open, close, adjusted close, low, high, volume, and date. The two datasets are linked through the ticker column.
Using Python, we will explore and understand the datasets before conducting more advanced analyses. Specifically, we will inspect the data, assess data quality, explore stock information, identify unusual observations, create visualizations, analyze stock prices over time, and identify opportunities for further analysis.

## **Part 1 — Load and Understand the Data**

### **Import libraries**

In [1]:
# Import libraries

import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

### **Load the datasets**

In [ ]:
# Load the datasets

stock_data = pd.read_csv("historical_stocks.csv")
stock_price_data = pd.read_csv("historical_stock_prices.csv")

### **Inspect to understand the historical_stock dataset**

In [ ]:
# Display the first five rows for the stock data

stock_data.head()

,ticker,exchange,name,sector,industry
0,PIH,NASDAQ,"1347 PROPERTY INSURANCE HOLDINGS, INC.",FINANCE,PROPERTY-CASUALTY INSURERS
1,PIHPP,NASDAQ,"1347 PROPERTY INSURANCE HOLDINGS, INC.",FINANCE,PROPERTY-CASUALTY INSURERS
2,TURN,NASDAQ,180 DEGREE CAPITAL CORP.,FINANCE,FINANCE/INVESTORS SERVICES
3,FLWS,NASDAQ,"1-800 FLOWERS.COM, INC.",CONSUMER SERVICES,OTHER SPECIALTY STORES
4,FCCY,NASDAQ,1ST CONSTITUTION BANCORP (NJ),FINANCE,SAVINGS INSTITUTIONS


In [14]:
# Inspect the datasets structure and dimension

stock_data.info()

<class 'pandas.DataFrame'>
RangeIndex: 6460 entries, 0 to 6459
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype
---  ------    --------------  -----
 0   ticker    6460 non-null   str  
 1   exchange  6460 non-null   str  
 2   name      6460 non-null   str  
 3   sector    5020 non-null   str  
 4   industry  5020 non-null   str  
dtypes: str(5)
memory usage: 640.9 KB


In [15]:
stock_data.shape

(6460, 5)

### **Inspect to understand the historical_stock_Price dataset**

In [6]:
# Display the first five rows for the stock price data

stock_price_data.head()

,ticker,open,close,adj_close,low,high,volume,date
0,AHH,11.50,11.58,8.493155,11.25,11.68,4633900,2013-05-08
1,AHH,11.66,11.55,8.471151,11.50,11.66,275800,2013-05-09
2,AHH,11.55,11.60,8.507822,11.50,11.60,277100,2013-05-10
3,AHH,11.63,11.65,8.544494,11.55,11.65,147400,2013-05-13
4,AHH,11.60,11.53,8.456484,11.50,11.60,184100,2013-05-14


In [11]:
# Find the company info for AHH

stock_data[stock_data["ticker"] == "AHH"]

,ticker,exchange,name,sector,industry
3543,AHH,NYSE,"ARMADA HOFFLER PROPERTIES, INC.",FINANCE,REAL ESTATE


In [ ]:
# Inspect the datasets structure and dimension

stock_price_data.info()

<class 'pandas.DataFrame'>
RangeIndex: 20973889 entries, 0 to 20973888
Data columns (total 8 columns):
 #   Column     Dtype  
---  ------     -----  
 0   ticker     str    
 1   open       float64
 2   close      float64
 3   adj_close  float64
 4   low        float64
 5   high       float64
 6   volume     int64  
 7   date       str    
dtypes: float64(5), int64(1), str(2)
memory usage: 1.5 GB


In [9]:
stock_price_data.shape

(20973889, 8)

### **Key insights from the dataset inspection:**

To be able to analyse our data, we loaded our datasets in the following way: Data from the "historical_stocks.csv" were stored in a variable called "stock_data" while data from "historical_stock_prices.csv" were stored in a variable called "stock_price_data". 

The historical_stocks dataset contains information about historical stock/securities. It includes the stock ticker (Identifier), the exchange where the stock is listed, the company name, and its sector and industry classifications. The ticker variable available also in the historical stock price will be used to link this dataset with the historical stock price dataset. The dataset has 6,460 records (rows) and 5 variables (Columns). All variables are stored as strings (Categorical). The ticker, exchange, and name columns have no missing values, while sector and industry each contain missing values. 

The historical_stock_prices dataset contains historical stock data. It has 20,973,889 records and 8 variables. It includes five continuous numerical variables (open, close, adj_close, low, and high), one integer variable (volume), and two string variables (ticker and date). The date variable is currently stored as a string and will be converted to a datetime format in the future steps for time-series analysis.

## **Part 2 — Check Data Quality**

### **Identify missing values**

In [ ]:
# Check for the number of missing values per column

print(stock_data.isnull().sum())

ticker         0
exchange       0
name           0
sector      1440
industry    1440
dtype: int64


In [23]:
# Percentage of missing values

print((stock_data.isnull().sum() / len(stock_data)) * 100)

ticker       0.000000
exchange     0.000000
name         0.000000
sector      22.291022
industry    22.291022
dtype: float64


In [ ]:
# Check for the number of missing values per column

print(stock_price_data.isnull().sum())

ticker       0
open         0
close        0
adj_close    0
low          0
high         0
volume       0
date         0
dtype: int64


In [ ]:
# Check/Understand the sector column

stock_data["sector"].value_counts(dropna=False)

sector
NaN                      1440
FINANCE                  1022
CONSUMER SERVICES         796
HEALTH CARE               784
TECHNOLOGY                607
CAPITAL GOODS             352
ENERGY                    286
PUBLIC UTILITIES          273
BASIC INDUSTRIES          272
CONSUMER NON-DURABLES     224
CONSUMER DURABLES         144
MISCELLANEOUS             139
TRANSPORTATION            120
SECTOR                      1
Name: count, dtype: int64

In [26]:
# Confirm whether the sector column has missing values

stock_data[stock_data["sector"].isnull()].head()

,ticker,exchange,name,sector,industry
19,ABP,NASDAQ,ABPRO CORPORATION,NaN,NaN
42,SQZZ,NASDAQ,ACTIVE ALTS CONTRARIAN ETF,NaN,NaN
62,ACT,NASDAQ,ADVISORSHARES VICE ETF,NaN,NaN
100,ABDC,NASDAQ,ALCENTRA CAPITAL CORP.,NaN,NaN
124,SMCP,NASDAQ,ALPHAMARK ACTIVELY MANAGED SMALL CAP ETF,NaN,NaN


In [28]:
# Are sector and industry missing together?

stock_data[["sector", "industry"]].isnull().sum()

sector      1440
industry    1440
dtype: int64

In [30]:
stock_data[stock_data["sector"].isnull() != stock_data["industry"].isnull()]

,ticker,exchange,name,sector,industry


In [25]:
# Check and understand the exchange column
stock_data["exchange"].value_counts(dropna=False)


exchange
NASDAQ    3308
NYSE      3152
Name: count, dtype: int64

#### **Insights on missing values:**
We identified 1,440 missing values in both the sector and industry variables, representing approximately 22.3% of the records. Since this is a big proportion of the dataset, removing these records could result in unnecessary loss of useful information about the stocks.

The sector column has 1440 missing (NaN) values and one class named "Sector" .................. (How to handle?????)


### **Check for duplicates**

In [21]:
# Check for dplicates

print(stock_data.duplicated().sum())


0


In [31]:
print(stock_price_data.duplicated(subset=["ticker", "date"]).sum())

0


#### **Insights on duplicates**

To check for duplicates in the stock price dataset, we created a combined index of ticker and date, since each ticker should have only one record for a given date. The analysis found no duplicate records. Similarly, the stock dataset was checked for duplicate tickers, and no duplicates were identified.